### PR TITLE
Issue #4383: add ignoreAnnotationElementDefaults property in MagicNumber

### PR DESCRIPTION
**Pull request by strkkk**
#4383

Reg report: https://strkkk.github.io/4383/

### CODE REVIEW

**Comment by strkkk** [comment]
CI is ok, one of travis builds failed with infra error

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.magicnumber;
+
+/**
+ * ignoreAnnotationElementDefaults = true/false
```
Comment (pbludov) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: Andrew, let me explain how this should work: 
One configuration == one test == one input file.

There should be simple
```
/**
 * ignoreAnnotationElementDefaults = true
 */
@interface InputMagicNumberAnnotationElement {
	int value() default 10; // OK
}
```
But the test `testDefault` should be extended to cover
```
@interface InputMagicNumberAnnotationElement {
	int value() default 10; // violation
}
```
if it does not have this case already.

Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: I thought about it, but decided to make a separate file for this option.
There is a one big file that is used for all tests, and I didn't want to add more examples there and correct all other test cases. It is difficult to maintain it, as for me.

Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: Corrected.

**Comment by strkkk** [comment]
@rnveach @romani please take a look

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java [hunk 2]
```diff
@@ -158,6 +166,22 @@
  * }
  * </pre>
  * <p>
+ * To configure the check allow annotation element defaults:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;MagicNumber&quot;&gt;
+ * &lt;property name=&quot;ignoreAnnotationElementDefaults&quot; value=&quot;true&quot;/&gt;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: Please indent property.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java [hunk 3]
```diff
@@ -279,6 +307,24 @@ public void visitToken(DetailAST ast) {
...
+        return ignoreAnnotation && isChildOf(ast, TokenTypes.ANNOTATION);
+    }
+
+    /**
+     * Checks if ast is annotation element default value and should be checked.
+     * @param ast token to check
+     * @return true if element is skipped, false otherwise
+     */
+    private boolean skipAnnotationDefaults(DetailAST ast) {
+        return ignoreAnnotationElementDefaults && isChildOf(ast, TokenTypes.LITERAL_DEFAULT);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: If you just check `LITERAL_DEFAULT` and no parent, won't it also ignore violations inside a `switch`'s `default` case?

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: No. In switch `default` literal is not parent of slist where magic number can appear, e.g.

```
~/Downloads $ javac A.java
~/Downloads $ cat A.java
class A {
  {   switch (1) {
            case 1:
            default: int b = 10;
      }
  }
}
~/Downloads $ java -jar checkstyle-8.22-all.jar -t A.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> A [1:6]
`--OBJBLOCK -> OBJBLOCK [1:8]
    |--LCURLY -> { [1:8]
    |--INSTANCE_INIT -> INSTANCE_INIT [2:2]
    |   `--SLIST -> { [2:2]
    |       |--LITERAL_SWITCH -> switch [2:6]
    |       |   |--LPAREN -> ( [2:13]
    |       |   |--EXPR -> EXPR [2:14]
    |       |   |   `--NUM_INT -> 1 [2:14]
    |       |   |--RPAREN -> ) [2:15]
    |       |   |--LCURLY -> { [2:17]
    |       |   |--CASE_GROUP -> CASE_GROUP [3:12]
    |       |   |   |--LITERAL_CASE -> case [3:12]
    |       |   |   |   |--EXPR -> EXPR [3:17]
    |       |   |   |   |   `--NUM_INT -> 1 [3:17]
    |       |   |   |   `--COLON -> : [3:18]
    |       |   |   |--LITERAL_DEFAULT -> default [4:12] <<<<< default
    |       |   |   |   `--COLON -> : [4:19]
    |       |   |   `--SLIST -> SLIST [4:21]
    |       |   |       |--VARIABLE_DEF -> VARIABLE_DEF [4:21]
    |       |   |       |   |--MODIFIERS -> MODIFIERS [4:21]
    |       |   |       |   |--TYPE -> TYPE [4:21]
    |       |   |       |   |   `--LITERAL_INT -> int [4:21]
    |       |   |       |   |--IDENT -> b [4:25]
    |       |   |       |   `--ASSIGN -> = [4:27]
    |       |   |       |       `--EXPR -> EXPR [4:29]
    |       |   |       |           `--NUM_INT -> 10 [4:29] <<<< number
    |       |   |       `--SEMI -> ; [4:31]
    |       |   `--RCURLY -> } [5:6]
    |       `--RCURLY -> } [6:2]
    `--RCURLY -> } [7:0]
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: > In switch default literal is not parent of slist where magic number can appear

Where is this `slist` check in the code?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java (hunk 1 above)
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: @pbludov If you are satisfied with the fix, please mark this conversation resolved.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 3 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: There is no need to check for slist, `isChildOf` is enough.

I updated test cases, please take a look.

**Comment by strkkk** [comment]
CI is ok, travis failed for infra issue

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * ignoreAnnotationElementDefaults = true
+ */
+@interface InputMagicNumberAnnotationElement {
+	int value() default 10; // OK
+	int[] value2() default {10, 20, 30}; // OK
+    class A {
+        {
+            switch (1) {
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: Add comment `// OK`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: @strkkk I looked at this in detail and it has nothing to do with `SLIST`.
`LITERAL_DEFAULT` is never the parent token for switch group, so it never gets picked up when scanning in `isChildOf`. `CASE_GROUP` is the parent token.
This is why the case isn't an issue.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 2 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: @strkkk This item came back.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.magicnumber;
+
+/**
+ * ignoreAnnotationElementDefaults = true
+ */
+@interface InputMagicNumberAnnotationElement {
+	int value() default 10; // OK
+	int[] value2() default {10, 20, 30}; // OK
+    class A {
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: corrected.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java (hunk 4 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberAnnotationElement.java]: added.

**Comment by rnveach** [comment]
@strkkk You may have missed the previous item at https://github.com/checkstyle/checkstyle/pull/6840#discussion_r296451643

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: corrected again.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java [hunk 6]
```diff
@@ -68,6 +68,11 @@
  * Default value is {@code false}.
  * </li>
  * <li>
+ * Property {@code ignoreAnnotationElementDefaults} -
+ * Ignore magic numbers in annotation elements defaults.
+ * Default value is {@code false}.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: from issue comment - https://github.com/checkstyle/checkstyle/issues/4383#issuecomment-504678505

> This Check need to skip such cases by default to my mind

please explain ignoring of this request.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java [hunk 7]
```diff
@@ -260,7 +287,8 @@ public MagicNumberCheck() {
 
     @Override
     public void visitToken(DetailAST ast) {
-        if ((!ignoreAnnotation || !isChildOf(ast, TokenTypes.ANNOTATION))
+        if (!skipAnnotationArgs(ast)
+                && !skipAnnotationDefaults(ast)
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: double negative in expressions ... can we avoid this ?

lets make it as:
```
 if (shouldTestAnnotationArgs(ast)
                && shouldTestAnnotationDefaults(ast)
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: Please check isChildOf(ast, TokenTypes.LITERAL_DEFAULT, TokenTypes.ANNOTATION);
to be exact, it is not a big a performace issue.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 6 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: This PR was opened before comment and I missed it.

I corrected code.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 7 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java (hunk 3 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java]: @romani sorry, I don't get your point.

**Comment by romani** [comment]
https://checkstyle.org/config_coding.html#MagicNumber_Properties look at `constantWaiverParentToken`

https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html look at ANNOTATION_FIELD_DEF

Can we simply suppress violation by usage of this token in property ?

**Comment by strkkk** [comment]
> Can we simply suppress violation by usage of this token in property ?

This property is an array of tokens. You mean add it into this list by default to skip this token branches?

It is possible, I think. In this case we don't need additional properties. To exclude this token, users should specify `constantWaiverParentToken` explicitly.

**Comment by romani** [comment]
@strkkk ,

> You mean add it into this list by default to skip this token branches?

yes.
the less properties we have the better. Extension of default will resolve issue in default Check configuration.
So users will get required results.
I think we should close this PR, and open new with extension of default value. If you agree - please do.
Input files and examples should be reused.
Sorry that I did not caught this on moment of approving the issue.

@rnveach  and @pbludov , fyi.

**Comment by strkkk** [comment]
@romani 
After some investigation of `constantWaiverParentToken` I saw that `LITERAL_DEFAULT` cannot be added to `constantWaiverParentToken` because it does not solve the problem. 
The thing is that tree looks like:
```
|--LITERAL_DEFAULT -> default [35:16]
|   `--EXPR -> EXPR [35:24]
|       `--NUM_INT -> 101 [35:24]
```
and like this for int array
```
|--LITERAL_DEFAULT -> default [36:15]
|   `--ANNOTATION_ARRAY_INIT -> { [36:23]
|       |--EXPR -> EXPR [36:24]
|       |   `--NUM_INT -> 102 [36:24]
|       `--RCURLY -> } [36:27]
```
for these two fields:
```
@interface AnnotationWithDefaultValue {
    int value() default 101;
    int[] ar() default {102};
}
```
and parent of number is always `EXPR`
I think that EXPR cannot be added as default token in `constantWaiverParentToken`
So problem is still relevant, I am reopening PR

**Comment by rnveach** [comment]
> constantWaiverParentToken

I don't understand the reason why the property doesn't work as [it looks like it should scan all parents](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java#L288-L300), but regardless, it can't be added in such a generic way as it will cause issues talked about at https://github.com/checkstyle/checkstyle/pull/6840#discussion_r296451799

**Pull request merged**